### ISSUE TITLE
Use fully qualifed domain name for storage backends when running on Kubernetes

### ISSUE THREAD

**Issue by electron0zero** [grafana/tempo#1726]
**Problem:**

Kubernetes configures `ndots=5` under `/etc/resolv.conf` options, that means any hostname that has less then 5 dots in it is looked up locally first, and then finally it is resolved.

As per our tests, we have seen 11 local lookups before DNS resolution is done (see https://github.com/grafana/jsonnet-libs/pull/762 for more details). These local lookups add latency, take up space in the conntrack table and add extra load to the kube-dns service. 

Tempo makes a tons of API calls to storage backend, and these extra DNS in storage backend calls adds to read and write latency for tempo.

We started looking into it because we noticed DNS Lookup Errors (see https://github.com/grafana/tempo/issues/1462), which we improved with https://github.com/grafana/tempo/pull/1632, and lowered ndots=3 to deal with it.

There are two ways to solve this: 
1. Use FQDN (hostname tld with dot) in storage backend config, and that will hint kube-dns to skip local search and do a direct lookup
2. lower ndots but it can break kubernetes service discovery [we can't do that]

solution 2 is not ideal so we are only left with solution 1.

This is NOT an issue if you are running tempo on VMs, In most Linux distros ndots is not configured,  and default value of 1 is used. if `cat /etc/resolv.conf  | grep ndots` shows nothing, that means it's set to default.

See more: 
- https://pracucci.com/kubernetes-dns-resolution-ndots-options-and-why-it-may-affect-application-performances.html
- https://github.com/grafana/jsonnet-libs/pull/762

**Solution:**

We start using FQDN in storage config, and suggest our users to use FQDN when running tempo on kubernetes.

**Azure Blob Storage:** If we just add dot at the end of TLD, we get 400 (see more https://github.com/grafana/tempo/issues/1462#issuecomment-1224082681), that happens because Azure doesn't allow trailing dot in `Host` header. We need to trim dot from Host header in Azure Client

**Google Cloud Storage:** not tested yet

**AWS S3:**  not tested yet

**Minio:** not tested yet

**Comment by AlexDHoffer** [comment]
Is the idea behind this to set all other configs aside from Azure to have a dot at the end of their base storage URLs? In other words, Azure will remain with a lot of DNS hits? Since looking at the Azure golang library one cannot set arbitrary request headers, and as such it doesn't seem possible using that library to get around the problem that they will not permit dots at the end of the container URL

**Comment by electron0zero** [comment]
I tested other providors, and looks like we will need to do what curl does with trailing dot.  curl trims traling dot from Host header but keeps it in URL, see https://github.com/curl/curl/commit/3a6142865fec3c54cd7081ada86c93c135e4b32f for details.

We can put trailing dot in config url, but trim it from `Host` header by using a custom transport in the client.

**Comment by Aki0x137** [comment]
@electron0zero I'd like to look into this issue/enhancement. Could you clarify the specific changes required?
Looking at the thread, I'm assuming expectation is to:

1. Add assertion to ensure that storage backend URIs use fully qualified domain names (FQDNs).
2. Append a trailing dot (`.`) at the end of the domain name( if not present) to ensure non-local DNS lookup.
3. When making API calls to the storage backend, strip the trailing dot from host header.

**Comment by electron0zero** [comment]
@Aki0x137 yup, we want to warn users on startup about it, if the backend URI is not FQDN. we want to have the same behaviour that curl has for handling this.

**Comment by knylander-grafana** [comment]
We could add an admonition to the K8s config docs that advises using FQDNs in storage backend configuration and point out the performance impact.

**Comment by electron0zero** [comment]
> [@electron0zero](https://github.com/electron0zero) is this a docs issue? is there anything actionable for us to change?

@joe-elliott to fully fix this, we need to add support for allowing trailing dot via a customer transport the the clients for object store, and send `Host` header with trailing dot stripped but keeps it in URL so we do DNS resolution with trailing dot.

docs are also good, but handling trailing dot in clients would be the most ideal solution here.

**Comment by priyendang** [comment]
Hi @electron0zero , I noticed this issue was previously being worked on but there hasn't been activity for a while. If it's still open, I'd be happy to take this up and work on a fix. Please let me know if that works.

**Comment by electron0zero** [comment]
all yours, feel free to tag me in PR for review.
